File: eco-core/core-nms/v1_17_R1/src/main/java/com/willfp/eco/proxy/v1_17_R1/fast/EcoFastItemStack.java
package com.willfp.eco.proxy.v1_17_R1.fast;

import com.willfp.eco.core.fast.FastItemStack;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.chat.ComponentSerializer;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.StringTag;
import net.minecraft.nbt.Tag;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.EnchantedBookItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import org.bukkit.craftbukkit.v1_17_R1.inventory.CraftItemStack;
import org.bukkit.craftbukkit.v1_17_R1.util.CraftChatMessage;
import org.bukkit.craftbukkit.v1_17_R1.util.CraftMagicNumbers;
import org.bukkit.craftbukkit.v1_17_R1.util.CraftNamespacedKey;
import org.bukkit.enchantments.Enchantment;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EcoFastItemStack implements FastItemStack {
    private final ItemStack handle;
    private final boolean isCIS;
    private final org.bukkit.inventory.ItemStack bukkit;
    private List<String> loreCache = null;

    public EcoFastItemStack(@NotNull final org.bukkit.inventory.ItemStack itemStack) {
        this.handle = FastItemStackUtils.getNMSStack(itemStack);
        if (itemStack instanceof CraftItemStack craftItemStack) {
            this.isCIS = true;
            this.bukkit = craftItemStack;
        } else {
            this.isCIS = false;
            this.bukkit = itemStack;
        }
    }

    @Override
    public Map<Enchantment, Integer> getEnchantmentsOnItem(final boolean checkStored) {
        ListTag enchantmentNBT = checkStored && handle.getItem() == Items.ENCHANTED_BOOK ? EnchantedBookItem.getEnchantments(handle) : handle.getEnchantmentTags();
        Map<Enchantment, Integer> foundEnchantments = new HashMap<>();

        for (Tag base : enchantmentNBT) {
            CompoundTag compound = (CompoundTag) base;
            String key = compound.getString("id");
            int level = '\uffff' & compound.getShort("lvl");

            Enchantment found = Enchantment.getByKey(CraftNamespacedKey.fromStringOrNull(key));
            if (found != null) {
                foundEnchantments.put(found, level);
            }
        }
        return foundEnchantments;
    }

    @Override
    public int getLevelOnItem(@NotNull final Enchantment enchantment,
                              final boolean checkStored) {
        ListTag enchantmentNBT = checkStored && handle.getItem() == Items.ENCHANTED_BOOK ? EnchantedBookItem.getEnchantments(handle) : handle.getEnchantmentTags();

        for (Tag base : enchantmentNBT) {
            CompoundTag compound = (CompoundTag) base;
            String key = compound.getString("id");
            if (!key.equals(enchantment.getKey().toString())) {
                continue;
            }

            return '\uffff' & compound.getShort("lvl");
        }
        return 0;
    }

    @Override
    public void setLore(@Nullable final List<String> lore) {
        loreCache = null;

        List<String> jsonLore = new ArrayList<>();
        if (lore != null) {
            for (String s : lore) {
                jsonLore.add(ComponentSerializer.toString(TextComponent.fromLegacyText(s)));
            }
        }

        CompoundTag displayTag = handle.getOrCreateTagElement("display");
        if (!displayTag.contains("Lore")) {
            displayTag.put("Lore", new ListTag());
        }

        ListTag loreTag = displayTag.getList("Lore", CraftMagicNumbers.NBT.TAG_STRING);
        loreTag.clear();
        for (String s : jsonLore) {
            loreTag.add(StringTag.valueOf(s));
        }

        apply();
    }

    @Override
    public List<String> getLore() {
        if (loreCache != null) {
            return loreCache;
        }

        List<String> lore = new ArrayList<>();

        for (String s : this.getLoreJSON()) {
            Component component = Component.Serializer.fromJson(s);
            lore.add(CraftChatMessage.fromComponent(component));
        }

        loreCache = lore;

        return lore;
    }

    private List<String> getLoreJSON() {
        CompoundTag displayTag = handle.getOrCreateTagElement("display");

        if (displayTag.contains("Lore")) {
            ListTag loreTag = displayTag.getList("Lore", CraftMagicNumbers.NBT.TAG_STRING);
            List<String> lore = new ArrayList<>(loreTag.size());
            for (int i = 0; i < loreTag.size(); i++) {
                lore.add(loreTag.getString(i));
            }
            return lore;
        } else {
            return new ArrayList<>();
        }
    }

    public void apply() {
        if (!this.isCIS) {
            bukkit.setItemMeta(CraftItemStack.asCraftMirror(handle).getItemMeta());
        }
    }
}

File: eco-core/core-nms/nms-common/src/main/kotlin/com/willfp/eco/internal/spigot/proxy/common/fast/EcoFastItemStack.kt
package com.willfp.eco.internal.spigot.proxy.common.fast

import com.willfp.eco.core.fast.FastItemStack
import com.willfp.eco.internal.spigot.proxy.common.commonsProvider
import com.willfp.eco.util.NamespacedKeyUtils
import com.willfp.eco.util.StringUtils
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.nbt.StringTag
import net.minecraft.world.item.EnchantedBookItem
import net.minecraft.world.item.Item
import net.minecraft.world.item.Items
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import kotlin.experimental.and

@Suppress("UsePropertyAccessSyntax")
class EcoFastItemStack(
    private val bukkit: org.bukkit.inventory.ItemStack
) : FastItemStack {
    private var loreCache: List<String>? = null
    private val handle = commonsProvider.asNMSStack(bukkit)

    override fun getEnchants(checkStored: Boolean): Map<Enchantment, Int> {
        val enchantmentNBT =
            if (checkStored && handle.getItem() === Items.ENCHANTED_BOOK) EnchantedBookItem.getEnchantments(
                handle
            ) else handle.getEnchantmentTags()
        val foundEnchantments: MutableMap<Enchantment, Int> = HashMap()
        for (base in enchantmentNBT) {
            val compound = base as CompoundTag
            val key = compound.getString("id")
            val level = ('\uffff'.code.toShort() and compound.getShort("lvl")).toInt()
            val found = Enchantment.getByKey(NamespacedKeyUtils.fromStringOrNull(key))
            if (found != null) {
                foundEnchantments[found] = level
            }
        }
        return foundEnchantments
    }

    override fun getLevelOnItem(
        enchantment: Enchantment,
        checkStored: Boolean
    ): Int {
        val enchantmentNBT =
            if (checkStored && handle.getItem() === Items.ENCHANTED_BOOK) EnchantedBookItem.getEnchantments(
                handle
            ) else handle.getEnchantmentTags()
        for (base in enchantmentNBT) {
            val compound = base as CompoundTag
            val key = compound.getString("id")
            if (key != enchantment.key.toString()) {
                continue
            }
            return ('\uffff'.code.toShort() and compound.getShort("lvl")).toInt()
        }
        return 0
    }

    override fun setLore(lore: List<String>?) {
        loreCache = null
        val jsonLore: MutableList<String> = ArrayList()

        if (lore != null) {
            for (s in lore) {
                jsonLore.add(StringUtils.legacyToJson(s))
            }
        }

        val displayTag = handle.getOrCreateTagElement("display")

        if (!displayTag.contains("Lore")) {
            displayTag.put("Lore", ListTag())
        }

        val loreTag = displayTag.getList("Lore", commonsProvider.nbtTagString)

        loreTag.clear()

        for (s in jsonLore) {
            loreTag.add(StringTag.valueOf(s))
        }

        apply()
    }

    override fun getLore(): List<String> {
        if (loreCache != null) {
            return loreCache!!
        }

        val lore = this.getLoreJSON().map { StringUtils.jsonToLegacy(it) }
        loreCache = lore
        return lore
    }

    private fun getLoreJSON(): List<String> {
        val displayTag = handle.getTagElement("display") ?: return emptyList()

        if (!displayTag.contains("Lore")) {
            return emptyList()
        }

        val loreTag = displayTag.getList("Lore", commonsProvider.nbtTagString)
        val lore = ArrayList<String>(loreTag.size)

        for (i in loreTag.indices) {
            lore.add(loreTag.getString(i))
        }

        return lore
    }

    override fun addItemFlags(vararg hideFlags: ItemFlag) {
        for (flag in hideFlags) {
            this.flagBits = this.flagBits or getBitModifier(flag)
        }

        apply()
    }

    override fun removeItemFlags(vararg hideFlags: ItemFlag) {
        for (flag in hideFlags) {
            this.flagBits = this.flagBits and getBitModifier(flag)
        }

        apply()
    }

    override fun getItemFlags(): MutableSet<ItemFlag> {
        val flags = mutableSetOf<ItemFlag>()

        var flagArr: Array<ItemFlag>
        val size = ItemFlag.values().also { flagArr = it }.size

        for (i in 0 until size) {
            val flag = flagArr[i]
            if (this.hasItemFlag(flag)) {
                flags.add(flag)
            }
        }

        return flags
    }

    override fun hasItemFlag(flag: ItemFlag): Boolean {
        val bitModifier = getBitModifier(flag)
        return this.flagBits and bitModifier == bitModifier
    }

    private var flagBits: Int
        get() =
            if (handle.hasTag() && handle.getTag()!!.contains(
                    "HideFlags",
                    99
                )
            ) handle.getTag()!!.getInt("HideFlags") else 0
        set(value) =
            handle.getOrCreateTag().putInt("HideFlags", value)

    override fun getRepairCost(): Int {
        return handle.getBaseRepairCost()
    }

    override fun setRepairCost(cost: Int) {
        handle.setRepairCost(cost)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is EcoFastItemStack) {
            return false
        }

        return other.hashCode() == this.hashCode()
    }

    override fun hashCode(): Int {
        return handle.getTag()?.hashCode() ?: (0b00010101 * 31 + Item.getId(handle.getItem()))
    }

    private fun apply() {
        commonsProvider.mergeIfNeeded(bukkit, handle)
    }

    private fun getBitModifier(hideFlag: ItemFlag): Int {
        return 1 shl hideFlag.ordinal
    }

    override fun unwrap(): org.bukkit.inventory.ItemStack {
        return bukkit
    }
}
